### PR TITLE
PLF-8746 : Perf problem when reading commentsId from an activity with…

### PR DESCRIPTION
**Pull request by rdenarie**
… a lot of comments

When getting commentId for an activity, we make 2 queries
- 1 to get comments from first level (activities where parentId = activityId)
- 1 to get subcomment (activities where parentId in (commentsId))

If an activity have a lot of comments (like space activity : all comments are member join)
the commentIds list can be huge.

When we make large user add in space, this particular request (with IN) is cached in SessionFactoryImpl.cacheQueryPlan. Each request with different list of id is one entry in cache.
In some case, it can fill the JVM memory.

We replace theses 2 requests by only one : we search activity where parentId=activityId of where parent.parentId=activityId
We search comments and subComments in one request, which can be cached once in cacheQueryPlan.

### CODE REVIEW

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java [hunk 1]
```diff
@@ -142,21 +142,20 @@ private ExoSocialActivity fillActivityFromEntity(ActivityEntity activityEntity,
     List<String> replyToIds = new ArrayList<>();
     List<ActivityEntity> comments = activityEntity.getComments() != null ? activityEntity.getComments()
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java]: This line is useless now.

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/entity/ActivityEntity.java [hunk 2]
```diff
@@ -48,6 +48,8 @@
         @NamedQuery(name = "SocActivity.findCommentsOfActivities", query = "SELECT a FROM SocActivity a "
             + " WHERE a.parent.id IN (:ids) "
             + " ORDER BY a.posted ASC"),
+        @NamedQuery(name = "SocActivity.findCommentsAndSubCommentOfActivity", query = "SELECT a FROM SocActivity a "
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/jpa/storage/entity/ActivityEntity.java]:
```suggestion
        @NamedQuery(name = "SocActivity.findCommentsAndSubCommentsOfActivity", query = "SELECT a FROM SocActivity a "
```

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java [hunk 3]
```diff
@@ -142,21 +142,20 @@ private ExoSocialActivity fillActivityFromEntity(ActivityEntity activityEntity,
...
-  private void fillCommentsIdsAndPosters(List<ActivityEntity> comments,
+  private void fillCommentsIdsAndPosters(ActivityEntity activity,
                                          List<String> commentPosterIds,
                                          List<String> replyToIds,
                                          boolean isSubComment) {
-    if (comments == null || comments.isEmpty()) {
-      return;
-    }
+    
+    List<ActivityEntity> comments = activityDAO.findCommentsAndSubCommentOfActivity(activity.getId());
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java]: It means the ActivityEntity already contains the comments (1st level), and we fetched them again (with the 2nd level), right ? Can't we avoid fetching 1st level comments twice ?

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java (hunk 1 above)
Comment (rdenarie) [reply component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java]: done

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/entity/ActivityEntity.java (hunk 2 above)
Comment (rdenarie) [reply component/core/src/main/java/org/exoplatform/social/core/jpa/storage/entity/ActivityEntity.java]: done

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java (hunk 3 above)
Comment (rdenarie) [reply component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java]: Comments are lazy loaded :
 @OneToMany(cascade = {CascadeType.REMOVE}, orphanRemoval=true, mappedBy="parent", fetch=FetchType.LAZY)
  private List<ActivityEntity> comments;
So, are you sure that they are already load at this time ?

Comment (thomasdelhomenie) [reply component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSActivityStorageImpl.java]: You are right, the call to getComments() has been removed, it should be ok then :+1:

**Pull request merged**